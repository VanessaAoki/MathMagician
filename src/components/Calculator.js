/* eslint-disable react/prefer-stateless-function */
/* eslint-disable react/no-array-index-key */
import React from 'react';
import calculate from '../logic/calculate';

class Calculator extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      total: null,
      next: null,
      operation: null,
    };
    this.onClickHandler = this.onClickHandler.bind(this);
  }

  onClickHandler(e) {
    this.setState((state) => calculate(state, e.target.innerHTML));
  }

  render() {
    const firstRow = ['AC', '+/-', '%', 'รท'];
    const secondRow = ['7', '8', '9', 'x'];
    const thirdRow = ['4', '5', '6', '-'];
    const fourthRow = ['1', '2', '3', '+'];
    const fifthRow = ['0', '.', '='];
    const { total, next, operation } = this.state;

    const first = firstRow.map(
      (elem, index) => (
        <button
          key={`first-${index}`}
          className={`first-${index}`}
          type="button"
          onClick={this.onClickHandler}
        >
          {elem}
        </button>
      ),
    );

    const second = secondRow.map(
      (elem, index) => (
        <button
          key={`second-${index}`}
          className={`second-${index}`}
          type="button"
          onClick={this.onClickHandler}
        >
          {elem}
        </button>
      ),
    );

    const third = thirdRow.map(
      (elem, index) => (
        <button
          key={`third-${index}`}
          className={`third-${index}`}
          type="button"
          onClick={this.onClickHandler}
        >
          {elem}
        </button>
      ),
    );

    const fourth = fourthRow.map(
      (elem, index) => (
        <button
          key={`fourth-${index}`}
          className={`fourth-${index}`}
          type="button"
          onClick={this.onClickHandler}
        >
          {elem}
        </button>
      ),
    );

    const fifth = fifthRow.map(
      (elem, index) => (
        <button
          key={`fifth-${index}`}
          className={`fifth-${index}`}
          type="button"
          onClick={this.onClickHandler}
        >
          {elem}
        </button>
      ),
    );
    return (
      <section className="calculator-grid">
        <div className="display-result">
          <p id="result">
            { next || total || '0' }
          </p>
          <p>
            { next == null ? operation : '' }
          </p>
        </div>
        <div className="first">{ first }</div>
        <div className="second">{ second }</div>
        <div className="third">{ third }</div>
        <div className="fourth">{ fourth }</div>
        <div className="fifth">{ fifth }</div>
      </section>
    );
  }
}

export default Calculator;
