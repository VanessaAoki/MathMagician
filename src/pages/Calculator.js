import React from 'react';

const Calculator = () => {
  <h2>Calculator</h2>;
};

export default Calculator;
