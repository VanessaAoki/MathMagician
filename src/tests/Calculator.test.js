import React from 'react';
import { render as rtlRender, screen } from '@testing-library/react';
import '@testing-library/jest-dom';
import userEvent from '@testing-library/user-event';
import renderer from 'react-test-renderer';
import Calculator from '../components/Calculator';

const render = (ui) => {
  window.history.pushState({}, 'Test page');

  return rtlRender(ui);
};

it('Calculator UI renders as expected', () => {
  const tree = renderer.create(<Calculator />).toJSON();
  expect(tree).toMatchSnapshot();
});

it('The calculator UI accepts the user\'s inputs and displays it, click on 4, then 3 and it should display 43', () => {
  render(<Calculator />);
  userEvent.click(screen.getByText(/4/i));
  userEvent.click(screen.getByText(/3/i));

  expect(screen.getByText(/43/i)).toBeInTheDocument();
});

it('The calculator UI and the logic are integrated, do 7 times 8 and it should display 56', () => {
  render(<Calculator />);
  userEvent.click(screen.getByText(/7/i));
  userEvent.click(screen.getByText(/X/i));
  userEvent.click(screen.getByText(/8/i));
  userEvent.click(screen.getByText(/=/i));

  expect(screen.getByText(/56/i)).toBeInTheDocument();
});
