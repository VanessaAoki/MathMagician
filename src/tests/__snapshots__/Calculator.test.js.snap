// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Calculator UI renders as expected 1`] = `
<section
  className="calculator-grid"
>
  <div
    className="display-result"
  >
    <p>
      0
    </p>
    <p />
  </div>
  <div
    className="first"
  >
    <button
      className="first-0"
      onClick={[Function]}
      type="button"
    >
      AC
    </button>
    <button
      className="first-1"
      onClick={[Function]}
      type="button"
    >
      +/-
    </button>
    <button
      className="first-2"
      onClick={[Function]}
      type="button"
    >
      %
    </button>
    <button
      className="first-3"
      onClick={[Function]}
      type="button"
    >
      รท
    </button>
  </div>
  <div
    className="second"
  >
    <button
      className="second-0"
      onClick={[Function]}
      type="button"
    >
      7
    </button>
    <button
      className="second-1"
      onClick={[Function]}
      type="button"
    >
      8
    </button>
    <button
      className="second-2"
      onClick={[Function]}
      type="button"
    >
      9
    </button>
    <button
      className="second-3"
      onClick={[Function]}
      type="button"
    >
      x
    </button>
  </div>
  <div
    className="third"
  >
    <button
      className="third-0"
      onClick={[Function]}
      type="button"
    >
      4
    </button>
    <button
      className="third-1"
      onClick={[Function]}
      type="button"
    >
      5
    </button>
    <button
      className="third-2"
      onClick={[Function]}
      type="button"
    >
      6
    </button>
    <button
      className="third-3"
      onClick={[Function]}
      type="button"
    >
      -
    </button>
  </div>
  <div
    className="fourth"
  >
    <button
      className="fourth-0"
      onClick={[Function]}
      type="button"
    >
      1
    </button>
    <button
      className="fourth-1"
      onClick={[Function]}
      type="button"
    >
      2
    </button>
    <button
      className="fourth-2"
      onClick={[Function]}
      type="button"
    >
      3
    </button>
    <button
      className="fourth-3"
      onClick={[Function]}
      type="button"
    >
      +
    </button>
  </div>
  <div
    className="fifth"
  >
    <button
      className="fifth-0"
      onClick={[Function]}
      type="button"
    >
      0
    </button>
    <button
      className="fifth-1"
      onClick={[Function]}
      type="button"
    >
      .
    </button>
    <button
      className="fifth-2"
      onClick={[Function]}
      type="button"
    >
      =
    </button>
  </div>
</section>
`;
